File: crypto/crypto.cpp
#include <cstdint>
#include <cstdio>

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

// -- from MAME
u16 rotate_left(u16 value, int n) {
    int aux = value>>(16-n);
    return ((value<<n)|aux)%0x10000;
}

u16 rotxor(u16 val, u16 xorval) {
    u16 res = val + rotate_left(val,2);

    res = rotate_left(res,4) ^ (res & (val ^ xorval));

    return res;
}

u32 cps3_mask(u32 address, u32 key1, u32 key2) {
    // // ignore all encryption
    // if (m_altEncryption == 2)
    // 	return 0;

    address ^= key1;

    u16 val = (address & 0xffff) ^ 0xffff;

    val = rotxor(val, key2 & 0xffff);

    val ^= (address >> 16) ^ 0xffff;

    val = rotxor(val, key2 >> 16);

    val ^= (address & 0xffff) ^ (key2 & 0xffff);

    return val | (val << 16);
}

int main(int argc, char* argv[]) {
    // 3s = sfiii3nr1
    // 0xa55432b4, 0x0c129981
    u32 key1 = 0xa55432b4;
    u32 key2 = 0x0c129981;

    // test vectors:
    for (u32 addr = 0; addr < 0x080000; addr++) {
        printf("%08x: mask = %08x\n", addr, cps3_mask(addr, key1, key2));
    }

    return 0;
}

File: cps3emu/src/crypto.rs
//! DECRYPTION LOGIC
//! thanks to Andreas Naive via MAME/FBNeo code.
//! written while drunk as fuck at [Nightclub](https://start.gg/nightclub)

fn rotate_left(val: u16, n: i32) -> u16 {
    let aux: u32 = (val >> (16 - n)) as u32;
    ((((val as u32) << n) | aux) % 0x10000) as u16
}

fn rotxor(val: u16, xorval: u16) -> u16 {
    let mut res: u16 = val.wrapping_add(rotate_left(val, 2));

    rotate_left(res, 4) ^ (res & (val ^ xorval))
}

pub fn cps3_kdf(addr: u32, key: [u32; 2]) -> u32 {
    let mut addr = addr;

    addr ^= key[0];
    let mut val: u16 = ((addr & 0xffff) as u16) ^ 0xffff;
    val = rotxor(val, (key[1] & 0xffff) as u16);
    val ^= ((addr >> 16) as u16) ^ 0xffff;
    val = rotxor(val, (key[1] >> 16) as u16);
    val ^= ((addr & 0xffff) as u16) ^ ((key[1] & 0xffff) as u16);

    (val as u32) | ((val as u32) << 16)
}

#[cfg(test)]
mod tests {
    use std::{path::{Path, PathBuf}, fs::{self, File}, io::{BufReader, BufRead}};
    use regex::Regex;

    use super::*;

    #[test]
    fn mame_golden_check_kdf() {
        let KEY: [u32; 2] = [0xa55432b4, 0x0c129981];
        let path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("test/data/kdf_a55432b4_0c129981");
        let rdr = BufReader::new(File::open(path).unwrap());

        let kdf_re = Regex::new(r"([0-9a-fA-F]+): mask = ([0-9a-fA-F]+)").unwrap();
        for line in rdr.lines().take(0x1000) { // only do a small subset for speed
            if let Ok(line) = line {
                let caps = kdf_re.captures(&line).unwrap();
                println!("caps={:?}", caps);
                let addr: u32 = u32::from_str_radix(&caps[1], 16).unwrap();
                let gold_mask: u32 = u32::from_str_radix(&caps[2], 16).unwrap();

                assert_eq!(gold_mask, cps3_kdf(addr, KEY));
            }
        }
    }
}
